# Customer Data Analysis and Visualisations - Part 1 (orders_products_customers_merged file)

### This script (Part 1) contains the following points:

#### 01. Import Libraries and Files

#### 02. Examining the df_analysis (orders_products_customers_merged.csv file) dataframe

#### 03. Data Wrangling and Consistency Checks on orders_products_customers_merged

* Check NaN values and change data types<br>
Change data types

#### 04. Data Analysis and Visualisations

##### Addressing PII (personally identifiable information) data in the dataframe

##### Comparing spending habits in geographic areas

##### Spending habits across the whole dataset based on orders (with duplicate user_id)

##### Spending habits based on individual user_id i.e. customers (no duplicates)

##### Creating 'exclusion_flag' and excluding Low-activity customers

#### 05. Exporting both the full dataframe 'df_analysis' and the active customers only dataframe 'df_analysis_active'

*Creation of flags and profiles carried out in Part 2*


***

## 01. Import Libraries and Files

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Assign the main project folder path to the variable path and view
path = r'/Users/elsaekevall/Jupyter_Notebook/Career_Foundry/06_2022_Instacart_Basket_Analysis/'
path

'/Users/elsaekevall/Jupyter_Notebook/Career_Foundry/06_2022_Instacart_Basket_Analysis/'

In [3]:
# Use the os.path.join() function to import the orders_products_merged.csv file as a pandas dataframe
df_analysis = pd.read_csv(os.path.join(path, '02_Data', '02_2_Prepared_Data', 'orders_products_customers_merged.csv'))

***

## 02. Examining the df_analysis (orders_products_customers_merged.csv file) dataframe

In [4]:
# View the first sixty rows of the dataframe df_analysis
df_analysis.head(60)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
1            1   2398795        1             2                   3   
2            2    473747        1             3                   3   
3            3   2254736        1             4                   4   
4            4    431534        1             5                   4   
5            5   3367565        1             6                   2   
6            6    550135        1             7                   1   
7            7   3108588        1             8                   1   
8            8   2295261        1             9                   1   
9            9   2550362        1            10                   4   
10          10   2968173       15            15                   1   
11          11   1870022       15            17                   2   
12          12   1911383       15            18                   2   
13          13   2715276       15            21                   1   
14          14    487368       15            22                   1   
15          15   2293453       19             2                   5   
16          16   1973799       19             5                   6   
17          17    532817       19             7                   4   
18          18   1573906       21            10                   3   
19          19   1593000       31            10                   3   
20          20   2231262       31            17                   3   
21          21   2580647       43             6                   4   
22          22   2187180       43             9                   4   
23          23   2497897       52             1                   1   
24          24   1318871       52             2                   1   
25          25   1261384       52             3                   1   
26          26    580568       52             5                   2   
27          27   2428073       52             7                   2   
28          28    342306       52             9                   1   
29          29   1520353       52            10                   2   
30          30   2542086       52            11                   2   
31          31    944694       52            12                   1   
32          32   1268191       52            15                   2   
33          33   1122089       52            17                   2   
34          34   1498922       52            18                   2   
35          35    180919       52            19                   3   
36          36   2695875       52            21                   5   
37          37   3205960       67             1                   3   
38          38    970922       67             2                   1   
39          39    596585       67             3                   1   
40          40   2037087       67             4                   2   
41          41   2474111       67             5                   3   
42          42   1568481       67             6                   3   
43          43   2048527       67             8                   2   
44          44   2389031       67             9                   2   
45          45   2378562       67            10                   1   
46          46   1515698       67            11                   1   
47          47   1928050       67            12                   2   
48          48   1949274       67            13                   1   
49          49    551869       67            14                   1   
50          50    768572       67            16                   4   
51          51    150528       67            18                   5   
52          52     29758       67            19                   0   
53          53   1558976       67            21                   1   
54          54    863974       67            22                   1   
55          55   2308399    

In [5]:
# View the shape of the dataframe df_analysis
df_analysis.shape

(32404859, 33)

In [6]:
# View the data types
df_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 33 columns):
 #   Column                            Dtype  
---  ------                            -----  
 0   Unnamed: 0                        int64  
 1   order_id                          int64  
 2   user_id                           int64  
 3   order_number                      int64  
 4   orders_day_of_week                int64  
 5   order_hour_of_day                 int64  
 6   days_since_prior_order            float64
 7   first_order                       bool   
 8   product_id                        int64  
 9   add_to_cart_order                 int64  
 10  reordered                         bool   
 11  product_name                      object 
 12  aisle_id                          int64  
 13  department_id                     int64  
 14  prices                            float64
 15  price_range_loc                   object 
 16  busiest_day                       

***

## 03. Data Wrangling and Consistency Checks on orders_products_customers_merged

In [7]:
# Drop the Unnamed: 0 and _merge columns which are not required for this analysis 
df_analysis.drop(columns = ['Unnamed: 0', '_merge'], inplace = True)
df_analysis

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True         196   
1                         7                    15.0        False         196   
2                        12                    21.0        False         196   
3                         7                    29.0        False         196   
4                        15                    28.0        False         196   
...                     ...                     ...          ...         ...   
32404854                 15                     1.0        False       43553   
32404855                 11                     3.0        False       43553   
32404856                  7                     NaN         True       42338   
32404857                  5                    15.0        False       42338   
32404858                 14                     NaN         True       28818   

          add_to_cart_order  reordered  ... spending_flag  \
0                         1      False  ...   Low spender   
1                         1       True  ...   Low spender   
2                         1       True  ...   Low spender   
3                         1       True  ...   Low spender   
4                         1       True  ...   Low spender   
...                     ...        ...  ...           ...   
32404854                  2       True  ...   Low spender   
32404855                  2       True  ...   Low spender   
32404856                  4      False  ...   Low spender   
32404857                 16       True  ...   Low spender   
32404858                  8      False  ...   Low spender   

          customer_median_prior_order_days         frequency_flag  gender  \
0                                     20.0       Regular Customer  Female   
1                                     20.0       Regular Customer  Female   
2                                     20.0       Regular Customer  Female   
3                                     20.0       Regular Customer  Female   
4                                     20.0       Regular Customer  Female   
...                                    ...                    ...     ...   
32404854                              10.0      Frequent customer  Female   
32404855                              10.0      Frequent customer  Female   
32404856                              22.0  Non frequent customer    Male   
32404857                              22.0  Non frequent customer    Male   
32404858                              14.0       Regular Customer    Male   

            US_state age date_joined no_dependents  family_status  income  
0            Alabama  31  2019-02-17             3        married   40423  
1            Alabama  31  2019-02-17             3        married   40423  
2            Alabama  31  2019-02-17             3        married   40423  
3            Alabama  31  2019-02-17             3        married   40423  
4            Alabama  31  2019-02-17             3        married   40423  
...              ...  ..         ...           ...            ...     ...  
32404854   Minnesota  48  2017-10-11             2        married   36899  
32

### Check NaN values and change data types

In [8]:
# Count NaN values in the dataframe
df_analysis.isnull().sum()

order_id                                  0
user_id                                   0
order_number                              0
orders_day_of_week                        0
order_hour_of_day                         0
days_since_prior_order              2076096
first_order                               0
product_id                                0
add_to_cart_order                         0
reordered                                 0
product_name                              0
aisle_id                                  0
department_id                             0
prices                                 5127
price_range_loc                           0
busiest_day                               0
Busiest_days                              0
busiest_period_of_the_day                 0
max_order                                 0
loyalty_flag                              0
customer_average_prices                   0
spending_flag                             0
customer_median_prior_order_days

### Change data types

In [9]:
# Change the data types of the 'user_id' to string, and 'gender', 'US_state', 'no_dependents' 'family_status' to category
df_analysis = df_analysis.astype({'user_id': str, 'gender': 'category', 'US_state': 'category', 'no_dependents': 'int16', 'family_status': 'category'})

In [10]:
# Change the 'date_joined' data type  to datetime
df_analysis['date_joined'] = pd.to_datetime(df_analysis['date_joined'])

In [11]:
# Change the age and income data types to lower storage capacities 
df_analysis['age'] = df_analysis['age'].astype('int16')
df_analysis['income'] = df_analysis['income'].astype('int32')

In [12]:
# Change the data types of the 'order_id', 'product_id', 'aisle_id' and 'department_id' columns to string (these variables are qualitative nominal) and reordered to boolean (this is a yes/no variable))
df_analysis = df_analysis.astype({'order_id': str, 'product_id': str, 'aisle_id': str, 'department_id': str, 'reordered': bool})

In [13]:
# Change the data types of the 'loyalty_flag', 'spending_flag', and 'frequency_flag' to cartegory (these values are ordinal categories)
df_analysis = df_analysis.astype({'price_range_loc': 'category', 'busiest_day': 'category', 'Busiest_days': 'category', 'busiest_period_of_the_day': 'category', 'loyalty_flag': 'category',  'spending_flag': 'category', 'frequency_flag': 'category'})

In [14]:
# Change the data types of the 'days_since_prior_order' and 'customer_median_prior_order_days' columns to Int32
df_analysis = df_analysis.astype({'order_number': 'int16', 'orders_day_of_week': 'int16', 'order_hour_of_day': 'int16', 'add_to_cart_order': 'int16', 'max_order': 'int16'})

In [15]:
# Change the data types of the 'days_since_prior_order' and 'customer_median_prior_order_days' columns to Int16
df_analysis = df_analysis.astype({'days_since_prior_order': 'Int16', 'customer_median_prior_order_days': 'Int16'})

IOStream.flush timed out


In [16]:
# Show the data types in the data frame and check the memory usage
df_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 31 columns):
 #   Column                            Dtype         
---  ------                            -----         
 0   order_id                          object        
 1   user_id                           object        
 2   order_number                      int16         
 3   orders_day_of_week                int16         
 4   order_hour_of_day                 int16         
 5   days_since_prior_order            Int16         
 6   first_order                       bool          
 7   product_id                        object        
 8   add_to_cart_order                 int16         
 9   reordered                         bool          
 10  product_name                      object        
 11  aisle_id                          object        
 12  department_id                     object        
 13  prices                            float64       
 14  price_range_loc 

In [17]:
# View the shape of the dataframe df_analysis
df_analysis.shape

(32404859, 31)

***

## 04. Data Analysis and Visualisations

2. Consider any security implications that might exist for this new data. You’ll need to address any PII data in the data before continuing your analysis.

### Addressing PII (personally identifiable information) data in the dataframe

**Customer names are not required for this analysis (if at a later stage names are required they can be found using their user_id) and these columns were dropped previously. Other PII information such as date of birth, full address, email address, social security number, etc are not included. In most states it would be difficult to identify customers from only state, age, number of dependents and family status.**

3. The Instacart officers are interested in comparing customer behavior in different geographic areas. Create a regional segmentation of the data. You’ll need to create a “Region” column based on the “State” column from your customers data set.
Use the region information in this Wikipedia article to create your column (you only need to create regions, not divisions).<br>
Determine whether there’s a difference in spending habits between the different U.S. regions. (Hint: You can do this by crossing the variable you just created with the spending flag.)


### Comparing spending habits in geographic areas

In [18]:
# Code to divide states into regions

# Set up region lists based on https://simple.wikipedia.org/wiki/List_of_regions_of_the_United_States
northeast = ['Maine', 'New Hampshire', 'Vermont', 'Massachusetts', 'Rhode Island', 'Connecticut', 'New York', 'Pennsylvania', 'New Jersey']
midwest = ['Wisconsin', 'Michigan', 'Illinois', 'Indiana', 'Ohio', 'North Dakota', 'South Dakota', 'Nebraska', 'Kansas', 'Minnesota', 'Iowa', 'Missouri']
west = ['Idaho', 'Montana', 'Wyoming', 'Nevada', 'Utah', 'Colorado', 'Arizona', 'New Mexico', 'Alaska', 'Washington', 'Oregon', 'California', 'Hawaii']
south = ['Delaware', 'Maryland', 'District of Columbia', 'Virginia', 'West Virginia', 'North Carolina', 'South Carolina', 'Georgia', 'Florida', 'Kentucky', 'Tennessee', 'Mississippi', 'Alabama', 'Oklahoma', 'Texas', 'Arkansas', 'Louisiana']
 
# create empty lists
region = []  

for state in df_analysis['US_state']:  # for each state in the state column in the dataframe
    if state in northeast:
        region.append('Northeast')  # if it is in the northwest list add the variable 'Northeast' to the region column
    elif state in midwest:
        region.append('Midwest')    # or if it is in the midwest list add the variable 'Midwest' to the region column
    elif state in west:
        region.append('West')       # or if it is in the west list add the variable 'West' to the region column
    else:
        region.append('South')      # else add the variable 'South' to the region column

In [19]:
# create new column df_analysis['Region'] with the values in the region list and view first 60
df_analysis['region'] = region
df_analysis[['region', 'US_state']].head(60)

region   US_state
0     South    Alabama
1     South    Alabama
2     South    Alabama
3     South    Alabama
4     South    Alabama
5     South    Alabama
6     South    Alabama
7     South    Alabama
8     South    Alabama
9     South    Alabama
10  Midwest    Indiana
11  Midwest    Indiana
12  Midwest    Indiana
13  Midwest    Indiana
14  Midwest    Indiana
15    South  Louisiana
16    South  Louisiana
17    South  Louisiana
18     West    Montana
19    South   Oklahoma
20    South   Oklahoma
21    South  Tennessee
22    South  Tennessee
23    South    Alabama
24    South    Alabama
25    South    Alabama
26    South    Alabama
27    South    Alabama
28    South    Alabama
29    South    Alabama
30    South    Alabama
31    South    Alabama
32    South    Alabama
33    South    Alabama
34    South    Alabama
35    South    Alabama
36    South    Alabama
37  Midwest       Iowa
38  Midwest       Iowa
39  Midwest       Iowa
40  Midwest       Iowa
41  Midwest       Iowa
42  Midwest       Iowa
43  Midwest       Iowa
44  Midwest       Iowa
45  Midwest       Iowa
46  Midwest       Iowa
47  Midwest       Iowa
48  Midwest       Iowa
49  Midwest       Iowa
50  Midwest       Iowa
51  Midwest       Iowa
52  Midwest       Iowa
53  Midwest       Iowa
54  Midwest       Iowa
55  Midwest       Iowa
56  Midwest       Ohio
57  Midwest       Ohio
58    South   Oklahoma
59    South   Oklahoma

In [20]:
# Change the region data type to string
df_analysis['region'] = df_analysis['region'].astype(str)

In [21]:
# Show the data types in the data frame and check the memory usage
df_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 32 columns):
 #   Column                            Dtype         
---  ------                            -----         
 0   order_id                          object        
 1   user_id                           object        
 2   order_number                      int16         
 3   orders_day_of_week                int16         
 4   order_hour_of_day                 int16         
 5   days_since_prior_order            Int16         
 6   first_order                       bool          
 7   product_id                        object        
 8   add_to_cart_order                 int16         
 9   reordered                         bool          
 10  product_name                      object        
 11  aisle_id                          object        
 12  department_id                     object        
 13  prices                            float64       
 14  price_range_loc 

In [22]:
# Show the value counts for the regions
df_analysis['region'].value_counts(dropna=False)

South        10791885
West          8292913
Midwest       7597325
Northeast     5722736
Name: region, dtype: int64

In [23]:
# check the shape of the dataframe
df_analysis.shape

(32404859, 32)

### Spending habits across the whole dataset based on orders (with duplicate user_id)

In [24]:
# determine the difference in spending habits by using crosstab to compare spending_flag counts in each region
crosstab_spending_flag = pd.crosstab(df_analysis['region'], df_analysis['spending_flag'], dropna = False)
crosstab_spending_flag

spending_flag  High spender  Low spender
region                                  
Midwest              156154      7441171
Northeast            108245      5614491
South                210017     10581868
West                 160478      8132435

In [25]:
#Alternative method to determine the difference in spending habits 
df_crosstab_alt_all = df_analysis.groupby(['region', 'spending_flag'])['spending_flag'].count()
df_crosstab_alt_all

region     spending_flag
Midwest    High spender       156154
           Low spender       7441171
Northeast  High spender       108245
           Low spender       5614491
South      High spender       210017
           Low spender      10581868
West       High spender       160478
           Low spender       8132435
Name: spending_flag, dtype: int64

In [26]:
# Calculate the spending habits as percentages of the total number in the region (i.e. across the row)
df_analysis_per_all = round(df_crosstab_alt_all.groupby(level=0).apply(lambda x:100 * x / float(x.sum())), 2)
df_analysis_per_all

region     spending_flag
Midwest    High spender      2.06
           Low spender      97.94
Northeast  High spender      1.89
           Low spender      98.11
South      High spender      1.95
           Low spender      98.05
West       High spender      1.94
           Low spender      98.06
Name: spending_flag, dtype: float64

**Instacart customers in the South region place the most orders (10791885), followed by the West and Midwest regions, while the Northeast region has the least (5722736). Although the number of customers in the regions differ, the spending habits based on the number of ordes placed across the regions are similar with the proportion of high spenders (around 2%) and low spenders (around 98%). Low spenders are placing more orders.**

### Spending habits based on individual user_id i.e. customers (no duplicates)

In [27]:
# creating a dataframe without duplicate user_id
df_analysis_users = df_analysis.drop_duplicates('user_id', keep='last')

In [28]:
# check the shape of the dataframe
df_analysis_users.shape

(206209, 32)

In [29]:
# Show region counts
df_region_count = df_analysis_users['region'].value_counts()
df_region_count

South        68737
West         52565
Midwest      48519
Northeast    36388
Name: region, dtype: int64

In [30]:
# Determine the difference in spending habits of individual users by using crosstab to compare spending_flag counts in each region 
crosstab_spending_users = pd.crosstab(df_analysis_users['region'], df_analysis_users['spending_flag'], dropna = False)
crosstab_spending_users

spending_flag  High spender  Low spender
region                                  
Midwest                1267        47252
Northeast               884        35504
South                  1823        66914
West                   1396        51169

In [31]:
# Alternative method to determine the difference in spending habits of individual users
df_crosstab_alt_users = df_analysis_users.groupby(['region', 'spending_flag'])['spending_flag'].count()
df_crosstab_alt_users

region     spending_flag
Midwest    High spender      1267
           Low spender      47252
Northeast  High spender       884
           Low spender      35504
South      High spender      1823
           Low spender      66914
West       High spender      1396
           Low spender      51169
Name: spending_flag, dtype: int64

In [32]:
# Calculate the spending habits as percentages of the total number in the region (i.e. across the row)
df_analysis_per_users = round(df_crosstab_alt_users.groupby(level=0).apply(lambda x:100 * x / float(x.sum())), 2)
df_analysis_per_users

region     spending_flag
Midwest    High spender      2.61
           Low spender      97.39
Northeast  High spender      2.43
           Low spender      97.57
South      High spender      2.65
           Low spender      97.35
West       High spender      2.66
           Low spender      97.34
Name: spending_flag, dtype: float64

**Most Instacart customers live in the South region (68737), followed by the West and Midwest regions, while the Northeast region has the least (36388). Although the number of customers in the regions differ, the spending habits across the regions are similar with the proportion of high spenders (around 3%) and low spenders (around 97%).**

4. The Instacart CFO isn’t interested in customers who don’t generate much revenue for the app. Create an exclusion flag for low-activity customers (customers with less than 5 orders) and exclude them from the data. Make sure you export this sample.

### Creating 'exclusion_flag' and excluding Low-activity customers

In [33]:
# Using loc statement to create an exclusion flag
df_analysis.loc[df_analysis['max_order'] < 5, 'exclusion_flag'] = 'Low-activity customer'
df_analysis.loc[df_analysis['max_order'] >= 5, 'exclusion_flag'] = 'Active customer'

In [34]:
# Checking frequency of exclusion_flag
df_analysis['exclusion_flag'].value_counts()

Active customer          30964564
Low-activity customer     1440295
Name: exclusion_flag, dtype: int64

In [35]:
# View the head of the df_analysis_active dataframe
df_analysis.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  first_order product_id  add_to_cart_order  \
0                    <NA>         True        196                  1   
1                      15        False        196                  1   
2                      21        False        196                  1   
3                      29        False        196                  1   
4                      28        False        196                  1   

   reordered  ...    frequency_flag  gender US_state  age date_joined  \
0      False  ...  Regular Customer  Female  Alabama   31  2019-02-17   
1       True  ...  Regular Customer  Female  Alabama   31  2019-02-17   
2       True  ...  Regular Customer  Female  Alabama   31  2019-02-17   
3       True  ...  Regular Customer  Female  Alabama   31  2019-02-17   
4       True  ...  Regular Customer  Female  Alabama   31  2019-02-17   

  no_dependents family_status income  region   exclusion_flag  
0             3       married  40423   South  Active customer  
1             3       married  40423   South  Active customer  
2             3       married  40423   South  Active customer  
3             3       married  40423   South  Active customer  
4             3       married  40423   South  Active customer  

[5 rows x 33 columns]

In [36]:
# Check the shape of the dataframe
df_analysis.shape

(32404859, 33)

In [37]:
# using crosstab to compare exclsion_flag counts with first orders
crosstab_exclusion_flag = pd.crosstab(df_analysis['first_order'], df_analysis['exclusion_flag'], dropna = False)
crosstab_exclusion_flag

exclusion_flag  Active customer  Low-activity customer
first_order                                           
False                  29303281                1025482
True                    1661283                 414813

**The low activity customers are not all first order customers**

In [38]:
# Create a copy (avoid warning errors) of the new dataframe for the Active customer entries in the dataframe and view the dataframe
df_analysis_active = df_analysis.groupby('exclusion_flag').get_group('Active customer')
df_analysis_active = df_analysis_active.copy(deep = True)
df_analysis_active

order_id user_id  order_number  orders_day_of_week  \
0         2539329       1             1                   2   
1         2398795       1             2                   3   
2          473747       1             3                   3   
3         2254736       1             4                   4   
4          431534       1             5                   4   
...           ...     ...           ...                 ...   
32404852  2329472  202557            15                   6   
32404853   694731  202557            16                   1   
32404854  1320836  202557            17                   2   
32404855    31526  202557            18                   5   
32404858  3093936  205420             1                   4   

          order_hour_of_day  days_since_prior_order  first_order product_id  \
0                         8                    <NA>         True        196   
1                         7                      15        False        196   
2                        12                      21        False        196   
3                         7                      29        False        196   
4                        15                      28        False        196   
...                     ...                     ...          ...        ...   
32404852                 12                       6        False      43553   
32404853                 14                       2        False      43553   
32404854                 15                       1        False      43553   
32404855                 11                       3        False      43553   
32404858                 14                    <NA>         True      28818   

          add_to_cart_order  reordered  ...     frequency_flag  gender  \
0                         1      False  ...   Regular Customer  Female   
1                         1       True  ...   Regular Customer  Female   
2                         1       True  ...   Regular Customer  Female   
3                         1       True  ...   Regular Customer  Female   
4                         1       True  ...   Regular Customer  Female   
...                     ...        ...  ...                ...     ...   
32404852                  2       True  ...  Frequent customer  Female   
32404853                  2       True  ...  Frequent customer  Female   
32404854                  2       True  ...  Frequent customer  Female   
32404855                  2       True  ...  Frequent customer  Female   
32404858                  8      False  ...   Regular Customer    Male   

           US_state  age date_joined no_dependents family_status  income  \
0           Alabama   31  2019-02-17             3       married   40423   
1           Alabama   31  2019-02-17             3       married   40423   
2           Alabama   31  2019-02-17             3       married   40423   
3           Alabama   31  2019-02-17             3       married   40423   
4           Alabama   31  2019-02-17             3       married   40423   
...             ...  ...         ...           ...           ...     ...   
32404852  Minnesota   48  2017-10-11             2       married   36899   
32404853  Minnesota   48  2017-10-11             2       married   36899   
32404854  Minnesota   48  2017-10-11             2       married   36899   
32404855  Minnesota   48  2017-10-11             2       married   36899   
32404858  Tennessee   71  2019-11-12             1       married  142822   

           region   exclusion_flag  
0           South  Active customer  
1           South  Active customer  
2           South  Active customer  
3           South  Active customer  
4           South  Active customer  
...           ...              ...  
32404852  Midwest  Active customer  
32404853  Midwest  Active customer  
32404854  Midwest  Active customer  
32404855  Midwest  Active customer  
32404858    South  Active customer  

[30964564 rows x 33 columns]

In [39]:
# Checking frequency of exclusion_flag
df_analysis_active['exclusion_flag'].value_counts()

Active customer    30964564
Name: exclusion_flag, dtype: int64

In [40]:
# Change the exclusion_flag data type to category
df_analysis_active['exclusion_flag'] = df_analysis_active['exclusion_flag'].astype('category')

In [41]:
# Show the data types in the data frame and check the memory usage
df_analysis_active.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30964564 entries, 0 to 32404858
Data columns (total 33 columns):
 #   Column                            Dtype         
---  ------                            -----         
 0   order_id                          object        
 1   user_id                           object        
 2   order_number                      int16         
 3   orders_day_of_week                int16         
 4   order_hour_of_day                 int16         
 5   days_since_prior_order            Int16         
 6   first_order                       bool          
 7   product_id                        object        
 8   add_to_cart_order                 int16         
 9   reordered                         bool          
 10  product_name                      object        
 11  aisle_id                          object        
 12  department_id                     object        
 13  prices                            float64       
 14  price_range_loc 

In [42]:
# Check the shape of the dataframe
df_analysis_active.shape

(30964564, 33)

***

## 05.  Exporting both the full dataframe 'df_analysis' and the active customers only dataframe 'df_analysis_active'

In [43]:
# Export the df_analysis_active dataframe in csv (for the client)
df_analysis_active.to_csv(os.path.join(path, '02_Data', '02_2_Prepared_Data', 'analysis_active_customers.csv'))

In [44]:
# Export the df_analysis dataframe in csv format
df_analysis.to_csv(os.path.join(path, '02_Data', '02_2_Prepared_Data', 'analysis_all_customers.csv'))

*Analysis from this point onwards carried out only on the active customers dataframe df_analysis_active*<br>
*Creation of flags and profiles carried out in Part 2*

***